### PR TITLE
Custom background & foreground style for individual loggers

### PR DESCRIPTION
**Pull request by Overflwn**
This (partly?) implements #35 by allowing to map a background/foreground `QBrush` for the given logger name.

Note that I probably have to change the name "[...]Mapping" to something like "[...]Brush" as I don't think that the word "mapping" fits here.

One problem I still have:
The new style does not update unless I click on the logger window. I currently only tested that via the sample program, adding two buttons to the toolbar (which opens as a seperate window for me). I don't know what happens when I already have focus on the main logger window.
Any suggestions?

### CODE REVIEW

File: include/qspdlog/qspdlog.hpp [hunk 1]
```diff
@@ -105,6 +105,12 @@ class QSpdLog : public QWidget
      */
     std::optional<std::size_t> getMaxEntries() const;
 
+    void setForegroundMapping(const std::string& loggerName, std::optional<QBrush> brush);
+    std::optional<QBrush> getForegroundMapping(const std::string& loggerName) const;
+
+    void setBackgroundMapping(const std::string& loggerName, std::optional<QBrush> brush);
+    std::optional<QBrush> getBackgroundMapping(const std::string& loggerName) const;
```
Comment (arsdever) [inline include/qspdlog/qspdlog.hpp:112]: Isn't `(set/get)Logger(Background/Foreground)` better naming?

File: sample/sample.cpp [hunk 2]
```diff
@@ -70,6 +70,30 @@ void configureToolbar(
...
+    clearRedBackgroundThread1->connect(
+        clearRedBackgroundThread1,
+        &QAction::triggered,
+        [&logView](bool) {
+        logView.setForegroundMapping(
+            "thread 1",
+            std::nullopt
+        );
+        }
+    );
```
Comment (arsdever) [inline sample/sample.cpp:96]: I thought to allow the UI user to be able to set the color(s). This is the dev usage of the feature, IMO.

File: src/qspdlog.cpp [hunk 3]
```diff
@@ -196,3 +196,27 @@ std::optional<std::size_t> QSpdLog::getMaxEntries() const
...
+)
+{
+    _sourceModel->setBackgroundMapping(loggerName, brush);
+}
+std::optional<QBrush> QSpdLog::getBackgroundMapping(
+    const std::string& loggerName
+) const
+{
+    return _sourceModel->getBackgroundMapping(loggerName);
+}
```
Comment (arsdever) [inline src/qspdlog.cpp:222]: Seems you haven't configured your formatter. Please make sure you have `clang-format` installed and you run it on the source files before committing. I'll be introducing a git hook to do it for you in one of my upcoming PRs.

File: src/qspdlog_model.cpp [hunk 4]
```diff
@@ -145,3 +160,39 @@ QVariant QSpdLogModel::headerData(
 
     return QVariant();
 }
+void QSpdLogModel::setForegroundMapping(
+    const std::string& loggerName, std::optional<QBrush> brush
+)
+{
+    if(brush.has_value())
```
Comment (arsdever) [inline src/qspdlog_model.cpp:167]: You could also use the following notation with the `std::optional`, though it's also fine, and maybe much better the way you wrote.
```suggestion
    if(brush)
```

File: test/test_qspdlog.cpp [hunk 5]
```diff
@@ -145,6 +145,20 @@ private slots:
         QCOMPARE(widget.itemsCount(), 20);
     }
 
+    void backgroundForegroundColorTest()
+    {
+        QSpdLog widget;
+        QCOMPARE(widget.getBackgroundMapping("test"), std::nullopt);
+        widget.setBackgroundMapping("test", QBrush(Qt::red));
+        QCOMPARE(widget.getBackgroundMapping("test"), QBrush(Qt::red));
+        QCOMPARE(widget.getBackgroundMapping("test2"), std::nullopt);
```
Comment (arsdever) [inline test/test_qspdlog.cpp:154]: You are setting the value for the `test2` logger, that's why the `erase` case is not covered.

File: src/qspdlog_model.cpp [hunk 6]
```diff
@@ -145,3 +160,39 @@ QVariant QSpdLogModel::headerData(
...
+    const std::string& loggerName
+) const
+{
+    if(_foregroundMappings.contains(loggerName))
+        return _foregroundMappings.at(loggerName);
+
+    return std::nullopt;
+}
+void QSpdLogModel::setBackgroundMapping(
+    const std::string& loggerName, std::optional<QBrush> brush
```
Comment (arsdever) [inline src/qspdlog_model.cpp:182]: `const std::string&` -> `std::string_view`
```suggestion
    std::string_view loggerName, std::optional<QBrush> brush
```

File: src/qspdlog_model.cpp [hunk 7]
```diff
@@ -145,3 +160,39 @@ QVariant QSpdLogModel::headerData(
...
+)
+{
+    if(brush.has_value())
+        _foregroundMappings[loggerName] = brush.value();
+    else if(_foregroundMappings.contains(loggerName))
+        _foregroundMappings.erase(loggerName);
+}
+std::optional<QBrush> QSpdLogModel::getForegroundMapping(
+    const std::string& loggerName
```
Comment (arsdever) [inline src/qspdlog_model.cpp:173]: `const std::string&` -> `std::string_view`
```suggestion
    std::string_view loggerName
```

File: src/qspdlog_model.cpp [hunk 8]
```diff
@@ -145,3 +160,39 @@ QVariant QSpdLogModel::headerData(
 
     return QVariant();
 }
+void QSpdLogModel::setForegroundMapping(
+    const std::string& loggerName, std::optional<QBrush> brush
```
Comment (arsdever) [inline src/qspdlog_model.cpp:164]: `const std::string&` -> `std::string_view`
```suggestion
    std::string_view loggerName, std::optional<QBrush> brush
```

File: src/qspdlog_model.cpp [hunk 9]
```diff
@@ -145,3 +160,39 @@ QVariant QSpdLogModel::headerData(
...
+)
+{
+    if(brush.has_value())
+        _backgroundMappings[loggerName] = brush.value();
+    else if(_backgroundMappings.contains(loggerName))
+        _backgroundMappings.erase(loggerName);
+}
+std::optional<QBrush> QSpdLogModel::getBackgroundMapping(
+    const std::string& loggerName
```
Comment (arsdever) [inline src/qspdlog_model.cpp:191]: `const std::string&` -> `std::string_view`
```suggestion
    std::string_view loggerName
```

File: src/qspdlog_model.hpp [hunk 10]
```diff
@@ -26,6 +26,12 @@ class QSpdLogModel : public QAbstractListModel
     void setMaxEntries(std::optional<std::size_t> maxEntries);
     std::optional<std::size_t> getMaxEntries() const;
 
+    void setForegroundMapping(const std::string& loggerName, std::optional<QBrush> brush);
+    std::optional<QBrush> getForegroundMapping(const std::string& loggerName) const;
+
+    void setBackgroundMapping(const std::string& loggerName, std::optional<QBrush> brush);
+    std::optional<QBrush> getBackgroundMapping(const std::string& loggerName) const;
```
Comment (arsdever) [inline src/qspdlog_model.hpp:33]: `const std::string&` -> `std::string_view`
```suggestion
    void setForegroundMapping(std::string_view loggerName, std::optional<QBrush> brush);
    std::optional<QBrush> getForegroundMapping(std::string_view loggerName) const;

    void setBackgroundMapping(std::string_view loggerName, std::optional<QBrush> brush);
    std::optional<QBrush> getBackgroundMapping(std::string_view loggerName) const;
```

File: src/qspdlog.cpp (hunk 3 above)
Comment (arsdever) [inline src/qspdlog.cpp:222]: The same `const std::string&` -> `std::string_view` issue here.

File: include/qspdlog/qspdlog.hpp (hunk 1 above)
Comment (arsdever) [inline include/qspdlog/qspdlog.hpp:112]: `const std::string&` -> `std::string_view`

Comment (arsdever) [inline include/qspdlog/qspdlog.hpp:112]: Documentation is missing.

**Review by arsdever — CHANGES_REQUESTED** [review]

**Comment by arsdever** [comment]
> One problem I still have: The new style does not update unless I click on the logger window. I currently only tested that via the sample program, adding two buttons to the toolbar (which opens as a seperate window for me). I don't know what happens when I already have focus on the main logger window. Any suggestions?

Since you are modifying the underlying data, but not notifying the views about changes, they won't update. The reason you see them updated is not because of the data or some magic, but because the views get focused and they repaint/update, thus fetching the data from the model. You need to notify the model somehow about it being changed. See [`QAbstractItemModel::dataChanged`](https://doc.qt.io/qt-6/qabstractitemmodel.html#dataChanged)

**Comment by Overflwn** [comment]
> > One problem I still have: The new style does not update unless I click on the logger window. I currently only tested that via the sample program, adding two buttons to the toolbar (which opens as a seperate window for me). I don't know what happens when I already have focus on the main logger window. Any suggestions?
> 
> Since you are modifying the underlying data, but not notifying the views about changes, they won't update. The reason you see them updated is not because of the data or some magic, but because the views get focused and they repaint/update, thus fetching the data from the model. You need to notify the model somehow about it being changed. See [`QAbstractItemModel::dataChanged`](https://doc.qt.io/qt-6/qabstractitemmodel.html#dataChanged)

Aah okay, I'll try that later.

File: test/test_qspdlog.cpp (hunk 5 above)
Comment (Overflwn) [reply test/test_qspdlog.cpp:154]: Thanks

File: sample/sample.cpp (hunk 2 above)
Comment (Overflwn) [reply sample/sample.cpp:96]: Do you mean via context menus or something like that?

File: include/qspdlog/qspdlog.hpp (hunk 1 above)
Comment (Overflwn) [reply include/qspdlog/qspdlog.hpp:112]: Yes, I just quickly programmed this out and didn't rename it yet.

File: sample/sample.cpp (hunk 2 above)
Comment (arsdever) [reply sample/sample.cpp]: Well, I think the context menu can be an option. At least for now. We need some kind of color picker for this. Maybe for now just give an action to set a color, which opens a dialog for inputting the logger name and the colors (foreground and background, both in hex code `#rgba` format). Note, we shouldn't forget about the initialization of the color fields. 

What do you think about this?

File: src/qspdlog_model.hpp [hunk 11]
```diff
@@ -46,5 +46,5 @@ class QSpdLogModel : public QAbstractListModel
     std::deque<entry_t> _items;
     std::optional<std::size_t> _maxEntries;
     std::map<std::string, QBrush> _backgroundMappings;
-    std::map<std::string, QBrush> _foregroundMappings;
+    std::map<std::string, QColor> _foregroundMappings;
```
Comment (Overflwn) [inline src/qspdlog_model.hpp:49]: Changing to a QColor directly disallows setting a QBrush style, though a style is a very nieche gimmick imo.

Comment (arsdever) [reply src/qspdlog_model.hpp:49]: The thing is, there is no brush for the foreground.

Comment (Overflwn) [reply src/qspdlog_model.hpp:49]: Not? I haven't tried it out but [the docs](https://doc.qt.io/qt-6/qt.html#ItemDataRole-enum) say it's a QBrush for Qt::ForegroundRole

Comment (Overflwn) [reply src/qspdlog_model.hpp:49]: Nevermind you're right, only the color works, style gets ignored.

**Comment by arsdever** [comment]
I'm on my way to writing some unit tests based on the latest changes I made. Don't worry about that.

**Comment by arsdever** [comment]
@Overflwn I changed the implementation slightly (compared to my previous changes). Please feel free to play with it and give me feedback. I also added tests for the new code, increasing coverage.

**Pull request merged**